Introduce Dekaf control-plane components

jshearer opened the pull request:
This PR builds ontop of https://github.com/estuary/flow/pull/1642, but is different enough that I didn't want to push to that PR after it was already reviewed. It is the first couple parts of https://github.com/estuary/flow/issues/1622. We have:

* The new endpoint type
  ``` yaml
      endpoint:
        dekaf:
          variant: some-dekaf-variant
          config:
            token: "foo"
  ```
* Dekaf implements a stripped down `materialize_unary` that is/will be responsible for validating these endpoint/resource configs. 
  * **TODO:** Figure out how to expose the schemars schema for `DekafConfig` and `DekafResourceConfig`. I'm pretty sure it looks like teaching `dekaf::materialize_unary` about the spec RPC, but not 100% sure.
    * Does this mean that we'll need to create some sort of fake `connector_tags` row and set it to discover/cache the schema? That's probably possible, but probably also won't just work OOTB.
  * It appears that there are two separate places where this needs to happen, so I did it in both. Is that true?
    * `agent::ProxyConnectors`
    * `runtime::Runtime`.
* New scenario tests that cover dekaf-type materializations, indirection of config, etc.
* Moved the `flowctl::Client` logic to a new crate: `flow_client` that both `flowctl` and `dekaf` can depend on
* A backwards-compatible update to the Dekaf service that allows you to pass in the name of a materialization as username, and the corresponding bearer token as a password
  * **TODO:** This is stubbed out at the point where we need to validate the token -- need to figure out how to actually fetch the latest built spec for a particular task name, extract its config, unseal it, etc.
 * Change Dekaf behavior if it's provided a task name/valid token
   * **TODO:** Metadata/discovery: Return bindings for this task instead of all collections a user has grants to
   * **TODO:** Schema registry: For `/subjects`, return materialization bindings instead of all collections
   * **TODO:** Self-sign data-plane access token in order to actually read data when not provided a refresh token 

--- 
I think the above TODOs should be resolved by a follow-up PR in order to get this new materialization type to the UI team ASAP.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/1665)
<!-- Reviewable:end -->

jshearer committed:
dekaf: Introduce `MaterializationEndpoint::Dekaf`

A materialization of this kind is purely descriptive, and cannot be directly started. It is used to convey the intent to expose a set of bindings through Dekaf, which will look it up to determine things like binding names and field selection/projections.

jshearer committed:
crates/activate: Dekaf materializations should not get shards or recovery journals, but they _should_ get ops stats+logs collections

jshearer committed:
schema: Update `flow.schema.json` to include `MaterializationEndpoint::Dekaf`

jshearer committed:
dekaf: Allow for indirecting `DekafConfig` if/when desired

jshearer committed:
dekaf: Introduce and validate `DekafResourceConfig` to configure a particular dekaf binding

jshearer committed:
temporarily remove `dekaf` dependency on `flowctl`

jshearer committed:
runtime: Delegate to Dekaf's `unary_materialize` when appropriate

jshearer committed:
Factor out `flow-client` from `flowctl`

Largely retain the same functionality as `flowctl::Client`, just in a new home so that it can be shared without introducing inconvenient dependencies

jshearer committed:
dekaf: Clean up and stub out location where bearer token will be checked

jshearer committed:
test: Update Dekaf tests

jshearer committed:
dekaf: Refresh token is not a JWT

jgraettinger commented on crates/validation/tests/scenario_tests.rs:
```diff
@@ -8,6 +8,126 @@ fn test_golden_all_visits() {
+}
+
+#[test]
+fn test_dekaf_materialization_indirect_config() {
```
tests over model shapes / parsing / config indirection should live in the `sources` crate.

`validation` crate tests focus on validation behaviors.

jgraettinger reviewed (commented):
Looking good! Some comments for you.

jgraettinger commented on crates/dekaf/src/connector.rs:
```diff
@@ -0,0 +1,73 @@
+pub struct DekafResourceConfig {
+    /// The exposed name of the topic that maps to this binding. This
+    /// will be exposed through the Kafka metadata/discovery APIs.
+    pub topic_name: String,
```
This comment is _not_ to say that having dekaf's unary_materialize() understand Spec isn't a good idea too, if you can get that plumbed through reasonably. It's just not a requirement.

jshearer replied:
> We'll likely want the x-table-name (?) annotation here so that the UI knows how to fill this out.
Huh, I didn't know that's how they got filled out. Does flowctl also do this clientside? I would have thought it'd be part of discovery, and was expecting to have to implement this in the Dekaf `unary_materialize` implementation. An annotation is certainly easier here though!

jgraettinger replied:
It's filled out by the `agent` and by the `UI`. I wouldn't assume that `flowctl raw discover` has correct handling for it, but don't know for sure.

jshearer replied:
Done -- I found `x-collection-name`, pretty sure that's the one

jshearer committed:
Some PR feedback

jshearer commented on crates/runtime/src/unary.rs:
```diff
@@ -20,8 +20,17 @@ impl<L: LogHandler> Runtime<L> {
     ) -> anyhow::Result<materialize::Response> {
-        let response = self.serve_materialize(unary_in(request)).boxed();
-        unary_out(response).await
+        match materialization_spec::ConnectorType::try_from(
```
I wasn't really sure where to put it, that's helpful thanks 👍

jshearer commented on crates/flow-client/src/client.rs:
```diff
@@ -46,21 +52,109 @@ impl Client {
+        }
+    }
+
+    pub fn pg_client(&self) -> postgrest::Postgrest {
```
> There's a down-side to re-building the client from scratch every time: we will no longer re-use the underlying connection pool, and will re-dial connections every time.

Riiight, yeah. So what if `Client` kept around a single `Postgrest` client that it hands out whenever somebody calls `pg_client()` (`postgrest::Client` is a thin wrapper around an `Arc<InnerClient>` so it can be shared easily), and which only gets recreated upon refresh. In fact, we could even check if a refresh is needed inside the call to `pg_client()` so you don't even need to think about it: any time you call `.pg_client()`, you'll get a client with valid access credentials.

Ideally I'd love to wrap `Postgrest` in a wrapper that checks and auto-refreshes its access token every time you make an API call. That way it could modify its own header map and not need to get re-created at all. But I think my proposed solution above is a decent compromise

jgraettinger replied:
I think it's far too much coupling between `Client`, which we should seek to keep very un-opinionated, with a highly opinionated refresh flow.

For example, refresh() will create a refresh token if it doesn't already have one, even if there's a perfectly valid access token available. Is that appropriate for flowctl? I think "yes", though even that is possibly debatable. Is it _always_ appropriate? Definitely not.

Or what if we introduce a new mechanism for refresh tokens? We easily could -- we're currently ripping out gateway_auth_token after all.

I don't think Client should know anything about the manner in which tokens are obtained, and that that business logic belongs in free function(s) that operate _on_ a Client.

jshearer replied:
Is [this](https://github.com/estuary/flow/pull/1665/commits/a69e6593469b8c04189fe3d67080c52a15f2831b) what you meant? I think I got a bit lost between `flow_client::Client` and `postgrest::Postgrest` when reading your comments, but this should:

* Move opinionated refresh logic to bare function to avoid coupling `flow_client::Client` to specific token refresh logic
* Create a single `postgrest::Postgrest` at `flow_client::Client` creation time and hand out clones of it with auth headers set to the current access token upon request. This should allow all users of `pg_client()` to share a connection pool, while still allowing `flow_client::Client` to hand out `postgrest::Postgrest` instances to all of the places that need them

jgraettinger replied:
Back up a step, and look at how flowctl on `master` updates it's configuration and then writes it out on exit. This change, as is, breaks that workflow because token refresh is treated as in interior detail of `flow_client::Client`, and I'm saying: it shouldn't be.

`flow_client::Client` should be immutable _exactly because_ of the kind of bug which currently exists in `flowctl` within this branch.

This is how I think it should be structured:
* `flow_client::Client`, once created, is immutable. We don't do any interior mutation, ever.
* Have a `flow_client::refresh_authorizations(client: &Client, access_token: Option<String>, refresh_token: Option<String>) -> anyhow::Result<(String, String)>` routine which takes an optional access and/or refresh token, and returns a guaranteed access and refresh token (or an error).
  * This routine should be the _only_ place we parse the access token and do an expiration check -- currently this logic is also still duplicated in flowctl/lib.rs.
  * Feel free to implement a further subroutine for getting an access token from a refresh token, if that's useful for dekaf.
* _Use_ this method in `flowctl` to update the configured access and refresh token, and to then re-build the client, as it had been doing prior to this PR.

IIRC there isn't a reason for a `flow_client::Client::claims()` routine aside from checking expiration, but I might be missing something. In any case, IMO this too should be a free function if it's still needed and Client should not have an opinion on the shape of an access token.

jshearer replied:
I see the problem, and why the raw access/refresh tokens need to be exposed -- in order for them to get set on `flowctl::Config` + written back to disk. I believe [this](https://github.com/estuary/flow/pull/1665/commits/f9fcd6cfe50e2004b398e4eb2b84092944742a3c) fixes `flowctl` writing updated creds back to disk. I tested it locally by removing `user_access_token` from my `local.json` profile, ran `cargo run -p flowctl auth roles list` and it wrote back a new access token

jgraettinger replied:
Yep, that works :+1:

jshearer commented:
Fixes https://github.com/estuary/flow/issues/1657

jshearer committed:
dekaf: Implement a couple memory optimizations

jshearer committed:
dekaf: Add some important annotations to endpoint/resource schemas

jshearer committed:
flow-client: Refactor to extract out refresh logic into bare function, and re-use Postgrest connection pool instead of creating a new one on each call to `pg_client()`

jshearer commented on crates/sources/src/loader.rs:
```diff
@@ -749,22 +764,30 @@ impl<F: Fetcher> Loader<F> {
-    async fn load_config<'s>(&'s self, scope: Scope<'s>, config: &RawValue) {
-        // If `config` is a JSON string that has no whitespace then presume and
-        // require that it's a relative or absolute URL to an imported file.
+    /// If `config` is a JSON string that has no whitespace then presume and
```
Good point, it was from when Dekaf stuff was in here. Will undo 👍

jshearer commented:
@jgraettinger re-requested your review. I believe I've addressed everything, except https://github.com/estuary/flow/pull/1665#discussion_r1785087662 if I didn't understand exactly what you wanted there.

jshearer committed:
dekaf: Always decrement `total_connections`, even if error

jshearer committed:
flow_client: Refactor to expose refreshed credentials

`flowctl` needs these credentials exposed in order to store them in `flowctl::Config` + on disk for the next flowctl invocation

jshearer committed:
dekaf: Expose more useful api histogram buckets

jshearer committed:
flowctl: Fix running unauthenticated

jshearer committed:
dekaf: Enable tuning connection pool size

jshearer committed:
dekaf: add more metrics and break out `dekaf_request_time` by request API key

jshearer committed:
dekaf: Frequently kill idle connections

jshearer committed:
dekaf: Remove connection healthcheck as `FindCoordinator` takes ~3-5 seconds to respond

jgraettinger reviewed (approved):
LGTM

jshearer committed:
dekaf: Add a bunch of descriptive timeouts

jshearer committed:
dekaf: Temporarily implement and use `flow_client::Client::with_fresh_gazette_client()`

gRPC connections seem to lose stability after a while, this temporarily fixes that

jshearer committed:
dekaf: Request clients to throttle if we hit a timeout
